### ISSUE TITLE
Erasure code guarantees are not kept under some circumstances

### ISSUE THREAD

**Issue by zerkms** [minio/minio#5825]
Scenario:

1. Create a 4 nodes cluster (m1, m2, m3, m4)
2. Shut down one node (m1)
3. Upload a file
4. Boot up a node (m1)
5. Shut down 2 more nodes (m2, m3)

Now - the data is lost.
The same scenario is reproducible in case if it was an intermittent network failure on a single node or if it was firewall mis-configuration or anything else.

I think it should be a command that checks that every node contains all data chunks it supposed to host, otherwise it's subject to silent data loss.

Using `mc admin heal` is not a solution, but an ugly workaround: you should not heal what's healthy.

Basically, if a node has lost write quorum - when it's reconnected to a cluster it should run the internal health check procedure and report correspondingly.

Your thoughts?

**Comment by fwessels** [comment]
In the exact steps that you describe you are correct that the data becomes inaccessible after step 5.

However, if we were to do this:
   6. boot either one of the down nodes

then the data would reappear, so it is not "irreversibly" lost (unless the disks on both m2 and m3 would crash simultaneously).

If you want to protect yourself against 2 down nodes, then we would advise to run a larger cluster of minimally 6 nodes or more (with the max being 32 nodes).

Minio is designed to sustain failures and not needing immediate repair (provided it is dimensioned according to the desired uptime guarantees). While "automatic" healing upon restarting a node may sound appealing, at large scale such scans can start to take hours or more, and as such we give control to users to either explicitly repair or not. This also allows eg. critical data to be prioritized over non-critical data, or do repairs at a time that is more appropriate (less busy).

Hopefully this addresses your concerns.

**Comment by zerkms** [comment]
> then the data would reappear, so it is not "irreversibly" lost (unless the disks on both m2 and m3 would crash simultaneously).
`s/Shut down/Destroy`

My point here is that neither it auto-heals itself, nor it reports that the state of the node is not consistent.

> Hopefully this addresses your concerns.

I have realised that it's what it is. The problem is that documentation does not focus on it. From the first glance it looks like the "erasure code" works great, while in fact - even 1 second node restart may cause the silent data loss eventually.

> we give control to users to either explicitly repair or not

there is another problem with it though: https://github.com/minio/minio/issues/5826

**Comment by fwessels** [comment]
> > then the data would reappear, so it is not "irreversibly" lost (unless the disks on both m2 and m3 would crash simultaneously).

> s/Shut down/Destroy
> My point here is that neither it auto-heals itself, nor it reports that the state of the node is not consistent.

Even auto healing upon startup would not always solve this due to the significant time (at scale) it would consume because presumably the destroy action would take place quickly after step 4.

> I have realised that it's what it is. The problem is that documentation does not focus on it. From the first glance it looks like the "erasure code" works great, while in fact - even 1 second node restart may cause the silent data loss eventually.

We recommend to run `mc admin heal` once in a while. And depending on the amount of nodes, stability of the network, parity setting, capacity of the cluster, HDD or SSD, etc. you can do so more or less frequently. 

> there is another problem with it though: #5826

We are looking into this.

**Comment by zerkms** [comment]
> because presumably the destroy action would take place quickly after step 4.

Nope, my original concern is that the node does not report about itself at all. So `Tstep5 - Tstep4` might be months.

> We recommend to run `mc admin heal` once in a while

Wouldn't `mc admin heal -r` take the same amount of resources (A LOT) as the auto-healing on start?

**Comment by fwessels** [comment]
Correct, these operations will tax the system considerably for a significant amount of time (provided a lot of data is under management).

Which is all the more reason why always automatically starting two heal operations in parallel when both m2 and m3 are restarted may not be what you would want as an operator (in addition to the regular production workload).

Of course if this is the desired behaviour then it is easy to eg. write a script that adds an `mc admin heal` after restarting the server.

**Comment by zerkms** [comment]
> Correct, these operations will tax the system considerably for a significant amount of time (provided a lot of data is under management).

What is the purpose of running it in cron as suggested, instead of doing it once on start, automatically? And given the nodes are in cluster - they might coordinate to do it one after another.

To me doing it once is better than doing it periodically.

**Comment by fwessels** [comment]
`mc admin heal` does more than just repair damage that may be caused because a node is down.

It will also detect any bit-rot that may have occurred over time and correct this. So it is advisable to run it on a recurring basis (say once per month) anyhow.

**Comment by zerkms** [comment]
"once a month" is an extremely large window, which significantly increases the chance of silent data loss.

**Comment by fwessels** [comment]
Not necessarily, for SSD-based configurations and/or larger numbers of nodes it may even be higher.

**Comment by deekoder** [comment]
I will close this issue and we can follow up  #5826 separately. Thanks @zerkms. Let us know if you have any further questions. You can also directly chat with us on our slack channel.
